--- a/net/minecraft/world/level/levelgen/NoiseChunk.java
+++ b/net/minecraft/world/level/levelgen/NoiseChunk.java
@@ -373,13 +_,24 @@
     }
 
     protected DensityFunction wrap(final DensityFunction function) {
-        return this.wrapped.computeIfAbsent(function, this::wrapNew);
+        // Canvas start - reduce worldgen allocations
+        // avoid lambda allocation
+        DensityFunction func = this.wrapped.get(function);
+
+        if (func == null) {
+            func = this.wrapNew(function);
+            this.wrapped.put(function, func);
+        }
+
+        return func;
+        // Canvas end - reduce worldgen allocations
     }
 
     private DensityFunction wrapNew(final DensityFunction function) {
         return switch (function) {
-            case DensityFunctions.Marker(DensityFunctions.Marker.Type type, DensityFunction wrapped) -> {
-                switch (type) {
+            case DensityFunctions.Marker marker -> {
+                final DensityFunction wrapped = marker.wrapped();
+                switch (marker.type()) {
                     case Interpolated:
                         yield new NoiseChunk.NoiseInterpolator(wrapped);
                     case FlatCache:
