--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
@@ -57,6 +_,21 @@
         levelChunk.getLevel().chunkPacketBlockController.modifyBlocks(this, chunkPacketInfo); // Paper - Anti-Xray - Modify blocks
     }
 
+    // Canvas start - extended view distance
+    public ClientboundLevelChunkWithLightPacket(
+        final int x,
+        final int z,
+        final ClientboundLevelChunkPacketData chunkData,
+        final ClientboundLightUpdatePacketData lightData
+    ) {
+        this.x = x;
+        this.z = z;
+        this.chunkData = chunkData;
+        this.lightData = lightData;
+        this.ready = true;
+    }
+    // Canvas end - extended view distance
+
     private ClientboundLevelChunkWithLightPacket(final RegistryFriendlyByteBuf input) {
         this.x = input.readInt();
         this.z = input.readInt();
