--- a/net/minecraft/server/network/PlayerChunkSender.java
+++ b/net/minecraft/server/network/PlayerChunkSender.java
@@ -81,7 +_,9 @@
     public static void sendChunk(final ServerGamePacketListenerImpl connection, final ServerLevel level, final LevelChunk chunk) { // Paper - rewrite chunk system - public
         // Paper start - Anti-Xray
         final boolean shouldModify = level.chunkPacketBlockController.shouldModify(connection.player, chunk);
-        connection.send(new ClientboundLevelChunkWithLightPacket(chunk, level.getLightEngine(), null, null, shouldModify));
+        final int hideY = io.canvasmc.canvas.chunk.antifreecam.AntiFreecam.hideBelowY(connection.player);
+        connection.send(new ClientboundLevelChunkWithLightPacket(chunk, level.getLightEngine(), null, null, shouldModify, hideY));
+        io.canvasmc.canvas.chunk.antifreecam.AntiFreecam.noteSent(connection.player, chunk.getPos(), hideY);
         // Paper end - Anti-Xray
         // Paper start - PlayerChunkLoadEvent
         if (io.papermc.paper.event.packet.PlayerChunkLoadEvent.getHandlerList().getRegisteredListeners().length > 0) {
